### ISSUE TITLE
Questions about restore

### ISSUE THREAD

**Issue by panda0307** [leela-zero/leela-zero#744]
In each iteration of self-play reinforcement learning, I have questions about optimization: 
1) Is it using the previous model instead of a new random initialized model? If it is so, the global_step is very large, isn't it? (e.g., 100,000 training steps in each optimization, and the step changes like 0-100,000-200,000-300,000-..., not 0-100,000)
2) Then, the number of training steps is fixed in each optimization. Is that true?
Thanks a lot!

**Comment by Alderi-Tokori** [comment]
1. Yes, it's using the best net as a starting point during each iteration of self-play. The global step is indeed getting larger and larger.
2. The number of training steps is somewhat fixed, snapshots are taken at arbitrarily decided step counts (in terms of the current iteration's step counts), and they are tested against the starting point, to see if they are stronger.

**Comment by panda0307** [comment]
@Alderi-Tokori , thanks. In AGZ paper, the number of training steps is 700,000? If it is so, it will waste a lot of time.

**Comment by panda0307** [comment]
The LZ uses different network architecture, but TensorFlow cannot restore a different model,how does it do with it?

**Comment by gcp** [comment]
First of all, we are not limited by the speed of the training, but by the speed with which the self-play games arrive.

Additionally, DeepMind had a fixed setup where they could tune learning rate, batching size and evaluation intervals against each other. We do not have that luxury. If it takes too long for the training games too arrive, the training machine will overtrain on the current dataset. This is why the reset to best network is done.

If you wanted to have something similar to the DM approach, my training machine would just sit idle most of the day until 25000 games are finished, and then do a training iteration, and send out a network. But your problem is now: if you continue training, you might have run far more training steps (too much!) than they would have before receiving a new batch of games. This is the calculation you have done above.

The problem is somewhat avoided by the approach of the second paper, where they just do a number of steps and send out the new network.

**Comment by gcp** [comment]
>The LZ uses different network architecture, but TensorFlow cannot restore a different model,how does it do with it?

I don't understand your question. The architecture is the same, just the amount of layers and filters differs.

**Comment by panda0307** [comment]
just the amount of layers and filters differs? Is that means **different** network structure?

**Comment by gcp** [comment]
Sorry, I don't understand what you're asking.

Are you asking how LZ went from 5x64 to 6x128? It's described in this thread: #591.

**Comment by panda0307** [comment]
In my opinion, the size of experience pool is fixed(e.g., 200,000 games), then, during each iteration of self-play, we add brand-new 25,000 games into the pool, at the same time, we drop the oldest games from the pool(also 25,000), and then we update the network parameters. I'm not sure whether I understand it correctly.

**Comment by gcp** [comment]
That is what AGZ did yes. But you don't know how many training steps are run in between the self-play batches.

**Comment by panda0307** [comment]
Yes, thanks, that's what I am saying. I will check it out in #591

**Comment by panda0307** [comment]
I don't understand the so-called 1000 checkpoint, could you please explain it to me? I would much appreciated with it.

**Comment by gcp** [comment]
I don't know what "1000 checkpoint" is either?

**Comment by panda0307** [comment]
Maybe, one iteration only has 1000 training steps? And the total number of training steps is 700,000?

**Comment by panda0307** [comment]
Yeah, the new network has been trained only after 8k training steps?

**Comment by gcp** [comment]
Current best perform network + 8000 steps.

Not 8000 steps from a random network.

**Comment by panda0307** [comment]
Oh, I see, it means 8k steps for updating NN'S parameters, during each iteration of self-play. Right?

**Comment by gcp** [comment]
There are no "iterations of self-play" because the clients are all independent and distributed. They send games continuously to a central server. The training machine continuously fetches the last 250k window, gets the best network, and starts training. It dumps new networks at 8k steps, 16k steps....until 128k steps. If one of these networks beats the old one by some margin, it becomes the best one. If none does, training just resets and we do another cycle (meanwhile more self-play games have arrived).

**Comment by panda0307** [comment]
With your experience, is "128k steps" be the best frequently?

**Comment by gcp** [comment]
No (or we'd have added 256k!).

**Comment by panda0307** [comment]
How about initialize the network by supervised learning, then reinforcement learning, after all, Master is also a good Go player! Maybe, it would reach high elo rating faster

**Comment by gcp** [comment]
It's a possibility, but a) for adding human knowledge inputs, it's a big question which ones are productive and which ones actually hurt in the long run b) seeing the knowledge that a player without human preconceptions develops is very interesting.

So it was simply chosen to not do this for this project.

**Comment by panda0307** [comment]
Thanks for your points!

**Comment by marcocalignano** [comment]
> No (or we'd have added 256k!).
I think we should explore even more than 256k since in the 6x128 there are more weights and it takes more steps to converge even with the same window.

**Comment by gcp** [comment]
It's not about getting full converge, it's about being able to improve by a minimal margin over the previous best.

**Comment by panda0307** [comment]
256k or 128k, is it depend on the size of experience pool?

**Comment by marcocalignano** [comment]
> It's not about getting full converge, it's about being able to improve by a minimal margin over the previous best.
I know but since your training machine has a lot of time I think you could change the termination condition from "I did 128K steps" to "my MSE is less that my goal" no matter how many steps it would take (with a limit of course if no condition is met)

**Comment by panda0307** [comment]
If you fetch less window, e.g., 10,000 instead of 25,000, so what? Is 25,000 necessary?

**Comment by jkiliani** [comment]
When you explain the reasons behind the backup to best-network, I'm wondering how successful the "Always adopt" strategy with Alpha Zero can be without some sort of mechanism to prevent overfitting. If you do decide to try this, how about making some sort of split between best-network (determined as before, and linked on the best-network list for people running bots etc.) and active-network, which is used for self-play. Would some sort of regular reset of the training to best-network  still be a good idea even with the Always-adopt technique?

**Comment by panda0307** [comment]
In AlphaZero paper, they say that the number of MCTS simulation is 800, is 800 too small？

**Comment by gcp** [comment]
>I'm wondering how successful the "Always adopt" strategy with Alpha Zero can be without some sort of mechanism to prevent overfitting.

I wonder too, but losing a ton of games and getting those training losses back in the next data set should correct the overfits a lot.

**Comment by Alderi-Tokori** [comment]
> Their mini-batch size is 8 times bigger than ours

I thought you were using a minibatch size of 512 (4 times smaller than AGZ)?

**Comment by gcp** [comment]
It's LZ=512. AGZ = 2048 and AZ = 4096.

**Comment by Alderi-Tokori** [comment]
I see, I hadn't realized they used a bigger minibatch size for AZ than for AGZ.

**Comment by sethtroisi** [comment]
This is now done.
